# お品書き
## 1. [NGSデータ解析チュートリアル](https://github.com/akihirao/how2cook/tree/main/ngs_training)
####  公開NGSデータの取得から変異検出までの解析フローを解説

## 2. [リードデータのDRAへの登録方法](https://github.com/akihirao/how2cook/tree/main/how2submit_DRA)
####  多検体のリードデータをDRAへ登録する方法の覚書

## 3. [ゲノムアッセンブリーのDDBJへの登録方法](https://github.com/akihirao/how2cook/tree/main/how2submit_assembly)
####  新規のゲノムアッセンブリーをMass Submission Ssytemを介してDDBJに登録する方法の覚書

## 4. [海底地形図の描写](https://github.com/akihirao/how2cook/blob/main/how2draw_ETOPO/Plot.Sea_around_JPN.md)
####  日本周辺の海底地形図をRで描写

## 5. VCF変換ツール
- [Vcf2PopCluster.pl](https://github.com/akihirao/how2cook/blob/main/Vcf2PopCluster.pl): A converter from VCF to PopCluster
- [vcf2newhybrid.py](https://github.com/akihirao/how2cook/blob/main/vcf2newhybs.py): A converter from VCF to NewHybrids
- [vcf2fasta_mt.py](https://github.com/akihirao/how2cook/blob/main/vcf2fasta_mt.py): A converter from VCF to multiple fasta for haploidy mitogenome
- [vcf2fasta_ref.py](https://github.com/akihirao/how2cook/blob/main/vcf2fasta_ref.py): A converter from VCF to fasta for haploidy reference genome
- [vcfgz2BA3.pl](https://github.com/akihirao/how2cook/blob/main/vcfgz2BA3.pl): A converter from VCF gz to BayesAss3 (BA3)
- [vcf2allele_depth_ratio.pl](https://github.com/akihirao/how2cook/blob/main/vcf2allele_depth_ratio.pl): Alternative allele depth ration in heterozygous genotypes


## 6. KMLマップ変換ツール
- [csv2kml.py](https://github.com/akihirao/how2cook/blob/main/csv2kml.py): A converter from coordinate csv to KML

## 7. [How to extract CDS from a GenBank format file](https://github.com/akihirao/how2cook/blob/main/how2extract_cds_from_gb/notes.md)
####  requirement: gbseqextractor https://github.com/linzhi2013/gbseqextractor

## 8. [Rパッケージ開発入門　10-12章　輪読会](https://github.com/akihirao/how2cook/blob/main/RpackDev/R_development_chap10-12.md)
